### PR TITLE
Generalize mbedtls_pk_setup_opaque to MBEDTLS_PSA_CRYPTO_CLIENT

### LINKED ISSUE #8710: Generalize mbedtls_pk_setup_opaque to MBEDTLS_PSA_CRYPTO_CLIENT
Extend `mbedtls_pk_setup_opaque` to provide support for `MBEDTLS_PK_OPAQUE` whenever PSA crypto API functions are available, i.e. whenever `MBEDTLS_PSA_CRYPTO_CLIENT` is enabled, regardless of the status of `MBEDTLS_USE_PSA_CRYPTO`.

For testing, we assume `MBEDTLS_PSA_CRYPTO_C`, since the case of `MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C` is currently untested and not officially supported.

See `docs/architecture/psa-migration/psa-legacy-bridges.md` (from https://github.com/Mbed-TLS/mbedtls/issues/7760) for some background.

Definition of done:

* The relevant library code and the declaration in `pk.h` are gated by `MBEDTLS_PSA_CRYPTO_CLIENT` instead of `MBEDTLS_USE_PSA_CRYPTO`.
* The unit tests for the function are enabled whenever `MBEDTLS_PSA_CRYPTO_C` is enabled.
* This may involve writing additional unit tests if the coverage is deemed insufficient according to the implementer and reviewers.

Related, but should in principle be independent: https://github.com/Mbed-TLS/mbedtls/issues/8712

### LINKED ISSUE #8712: Clarify the documentation of mbedtls_pk_setup_opaque
In the documentation of `mbedtls_pk_setup_opaque`, clarify which algorithm the functions `mbedtls_pk_{sign,verify,encrypt,decrypt}` will perform with the resulting key.

Definition of done: updated documentation, and possibly add unit tests if the documentation ends up stating things that are currently untested.

Related, but should in principle be independent: https://github.com/Mbed-TLS/mbedtls/issues/8710

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

This PR aims at:
- adding proper test components in `all.sh` for `CRYPTO_CLIENT`: one starting from `default` configuration and one starting from `full`.
- changing guards for `mbedtls_pk_setup_opaque` from `USE_PSA_CRYPTO` to `CRYPTO_CLIENT`.

Resolves #8710.
Resolves #8712.
Depends on #8774.

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required
- [x] **tests** provided (no new test added, but modified already existing ones)

### CODE REVIEW

File: library/pk_wrap.c
Comment (gilles-peskine-arm) [inline library/pk_wrap.c:525]: Functions that allow PK to cooperate with PSA should be guarded by `MBEDTLS_PSA_CRYPTO_CLIENT` (e.g. `mbedtls_pk_setup_opaque`, `mbedtls_pk_copy_from_psa`, `mbedtls_pk_import_into_psa`, …). But functions that cause PK to dispatch to PSA instead of built-in code should remain guarded by `MBEDTLS_USE_PSA_CRYPTO`: that's what it's for.

So the PSA version of `ecdsa_verify_wrap` needs to remain gated by `MBEDTLS_USE_PSA_CRYPTO`. But `ecdsa_opaque_verify_wrap` needs to be available whenever `MBEDTLS_PSA_CRYPTO_CLIENT` is defined.

File: tests/suites/test_suite_pk.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pk.data:547]: This pull request is not supposed to change the behavior in any tested configuration, so this change should not be needed. Also, since there is no change in the behavior underlying functions in `pk.c` in the non-PSA cases, this can't work when PSA is disabled. And indeed there are CI failures in components that disable PSA.

After investigation, the problem turns out to be that 7986c77bbd9442386132827adf9fd1aaeb298b32 is too eager and changes the dispatch in pk to PSA when it shouldn't. See my comments in `pk_wrap.c`.

File: include/mbedtls/pk.h
Comment (gilles-peskine-arm) [inline include/mbedtls/pk.h:391]: Now that `MBEDTLS_PK_OPAQUE` is enabled whenever `MBEDTLS_PSA_CRYPTO_CLIENT` is enabled, we need to test it. We can't test client-only configurations, but we still need to change the guards on testing of `MBEDTLS_PK_OPAQUE` from `MBEDTLS_USE_PSA_CRYPTO` to `MBEDTLS_PSA_CRYPTO_C`.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1279]: It bothers me that we're no longer running any testing with `MBEDTLS_PSA_CRYPTO_CLIENT && !MBEDTLS_PSA_CRYPTO_C`. The idea in https://github.com/Mbed-TLS/mbedtls/issues/8710#issuecomment-1933825279 was that we'd still run tests, just not doing any testing that invokes PSA functions. We do need to link programs that call PSA functions, but these functions can be stubs that just return an error. I've tried it any it works, at least, in the default config, in the full config and in the config of this component. https://github.com/gilles-peskine-arm/mbedtls/tree/issue8710-restore_non_psa_paths

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
This breaks configurations with PSA enabled but USE_PSA_CRYPTO enabled: in those configurations, pk should not dispatch to PSA. I think I have a fix, see my comments.

Also we need to test `MBEDTLS_PK_OPAQUE` whenever `MBEDTLS_PSA_CRYPTO_C` is enabled now.

I haven't done a complete review of which places are now guarded by what. I'll do that once the CI passes.

File: library/pk_wrap.c
Comment (valeriosetti) [reply library/pk_wrap.c:525]: Thanks for the explanation! This was exactly the doubt I had while working on this PR because `USE_PSA` and `CRYPTO_CLIENT` looked very very similar to me.

Comment (valeriosetti) [reply library/pk_wrap.c:525]: Last question. It seems to me that `mbedtls_pk_wrap_as_opaque()` might be guarded with `CRYPTO_CLIENT` as well for the same reason as `setup_opaque()`, but I guess that we're ignoring this because this is going to be deprecated, right?

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh]: Stubs are fine for the build, but then test with opaque keys which rely on PSA fails because all the stubbed functions do not work. I think this is the last failure (hopefully) I should fix now in the CI.

I will try to play a bit with `helper_libtestdriver1_make_drivers` and `helper_libtestdriver1_make_main` to see if I can get something that we can play with more easily.

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: > Stubs are fine for the build, but then test with opaque keys which rely on PSA fails because all the stubbed functions do not work.

Sure, the idea is that for now we build code with `CLIENT && !C` but we keep the tests depending on `C`. And that's only for the `!USE` build: with `USE_PSA_CRYPTO` pretty much all of pk would have to be excluded.

**Comment by valeriosetti** [comment]
Failures found in the OpenCI of the type `ABI-API-checking` are related to the fact that `priv_id` is now guarded by `CRYPTO_C` instead of `USE_PSA`.

File: tests/suites/test_suite_pk.function
Comment (mpg) [inline tests/suites/test_suite_pk.function:740]: I don't think that's correct. The function itself does not depend on `PSA_CRYPTO_C`. Only the parts that used to depend on `USE_PSA_CRYPTO` (and that you changed to `PSA_CRYPTO_CLIENT`) now depend on `PSA_CRYPTO_C`.

Also, we should change `USE_PSA_INIT/DONE` to `PSA_INIT/DONE()` (which do nothing if `PSA_CRYPTO_C` is not defined). This might apply to other test functions too.

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good, just a few things.

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:1321]: > And remove wrap_as_opaque since it's being removed.

Eheheh, yes, I definitely picked the wrong one :laughing:

File: tests/suites/test_suite_pk.function
Comment (valeriosetti) [reply tests/suites/test_suite_pk.function]: Yes, you're right. I also checked other functions I modified in this PR, but in all these cases I changed the overall function guard from `USE_PSA` to `CRYPTO_C` because those functions are heavily depending on opaque keys (not just a minor testing that can easily be excluded).
[This commit](https://github.com/Mbed-TLS/mbedtls/pull/8842/commits/fd4dcbfe8e428a5530121a215e4f822679aad5a2) should fix this problem :wink:

**Review by mpg — APPROVED** [review]
LGTM, thanks!

File: ChangeLog.d/8710.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/8710.txt:3]: Blocker: as of 4b2700e14934b815e73f9ef39a08705b732b7acc, the code in Mbed TLS that consumes pk keys does not correctly deal with the `MBEDTLS_PK_OPAQUE` case when `MBEDTLS_USE_PSA_CRYPTO` is disabled. I have not yet fully reviewed this aspect, but it's something we need to do.

At least one case that I've found while working on expanding testing: if you configure a TLS server for static ECDH with an opaque key, it crashes due to a null pointer dereference when it tries to use the key for ECDH. Found with the `Opaque key for server authentication: ECDH-` test of `ssl-opt.sh` after expanding the test programs. Backtrace:
```
0 in mbedtls_ecdh_get_params of source/library/ecdh.c:448
1 in ssl_get_ecdh_params_from_cert of source/library/ssl_tls12_server.c:2751
2 in ssl_write_server_key_exchange of source/library/ssl_tls12_server.c:3264
3 in mbedtls_ssl_handshake_server_step of source/library/ssl_tls12_server.c:4356
4 in mbedtls_ssl_handshake_step of source/library/ssl_tls.c:4022
5 in mbedtls_ssl_handshake of source/library/ssl_tls.c:4069
6 in main of source/programs/ssl/ssl_server2.c:3463
```
The code for call 1 is
```
    if ((ret = mbedtls_ecdh_get_params(&ssl->handshake->ecdh_ctx,
                                       mbedtls_pk_ec_ro(*mbedtls_ssl_own_key(ssl)),
                                       MBEDTLS_ECDH_OURS)) != 0) {
```
and `mbedtls_pk_ec_ro` returns `NULL` when the key is opaque.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Two major points, plus a few minor things:

* We need to ensure that we expand the test coverage whenever we expand the library. So far I'm happy with unit test coverage, but this needs to be carefully re-checked after merging with more functions in `pk.c` that are currently in flight. And I'm not happy with the TLS test coverage (and perhaps also X.509), which leads to the second point…
* There is a blocker bug in TLS (and perhaps also X.509) code, which doesn't accept the new kinds of PK objects.

Review done on 4b2700e14934b815e73f9ef39a08705b732b7acc

File: ChangeLog.d/8710.txt
Comment (gilles-peskine-arm) [reply ChangeLog.d/8710.txt:3]: Having dug further, I believe the only risk of a security hole or crash (as opposed to merely not functioning) are where the code calls one of the pk unsafe functions (`mbedtls_pk_rsa`, `mbedtls_pk_ec{,_ro,_rw}`). I've checked and I'm happy to report that the static ECDH case mentioned above is the only one. Making it work looks difficult, but I'm ok with releasing 3.6.0 with this as a documented limitation that causes an internal error at runtime.

Comment (gilles-peskine-arm) [reply ChangeLog.d/8710.txt:3]: https://github.com/Mbed-TLS/mbedtls/pull/8910 has passed the CI, so I recommend including it here. I believe it fully resolves this comment.

File: ChangeLog.d/8710.txt
Comment (mpg) [reply ChangeLog.d/8710.txt:2]: Also, it's only tested when `CRYPTO_C` is enabled, and we don't want to advertise untested features in the ChangeLog.

(As reminder, support for `CRYPTO_CLIENT && !CRYPTO_C` is entirely unofficial until we can test it.)

File: ChangeLog.d/8710.txt
Comment (mpg) [reply ChangeLog.d/8710.txt:3]: > If we claim it, we should test it.

Thanks noticing and raising 8910.

I'll repeat that I think for this reason, the ChangeLog should only mention `_C` not `_CLIENT`.

File: include/mbedtls/pk.h
Comment (mpg) [reply include/mbedtls/pk.h:227]: I was going to say we'd need an extra field to distinguish between the two, but actually no: the information is already there in `pk_info`. That is, if I'm not mistaken: if type is opaque then use `priv_id`, else if type is RSA (including alt) use `pk_ctx` else depends on `USE_PSA_EC_DATA`.

As you said, it's pre-existing, and it smells like a potential can of worms to me, so I'm inclined to not even try it here, and only consider it for a follow-up. (Perhaps not in 3.6.x though due to ABI concerns?)

Comment (gilles-peskine-arm) [reply include/mbedtls/pk.h:227]: Right. Definitely not try it here due to the risk, and can't do it in 3.6.x because it would change the ABI. I was asking because I was wondering if I was missing something. I find the arrangement of the “new” pk context layouts a little weird.

File: ChangeLog.d/8710.txt
Comment (mpg) [reply ChangeLog.d/8710.txt:3]: > Making it work looks difficult

My initial reaction when seeing #8905 was "this might be more difficult than you think", and then "is this really worth it?" Apparently for PK opaque you managed to make it work (I'm assuming, since 8910 is now passing CI). The PSK case may well turn out to be harder (I'm mostly worried about the mixed-PSK key exchanges RSA-PSK, (EC)DHE-PSK).

But I stand by my second point: why would anybody who cares about PSA in TLS not want to enable `USE_PSA_CRYPTO`? As discussed on Slack, that makes no sense to me.

> but I'm ok with releasing 3.6.0 with this as a documented limitation that causes an internal error at runtime.

That no longer needs to be considered since you made it work, but if it were I'd have argued against it. Having things that partially work impose an extra cognitive burden on our users and ourselves, not to mention making documenting and testing harder. And again, for very little gain in exchange: does anyone really want to use PSA crypto in TLS and not enable `USE_PSA_CRYPTO`? Why would anyone want that?

Currently the semantics of `USE_PSA_CRYPTO` is more complicated than it should be, for our user's sake as well as ours. On one hand, I can see how making the extra APIs like `setup_opaque` depend only on `_C` (or `_CLIENT` unofficially) could be a step towards making it simpler. On the other hand, from an implementor's perspective, I think it makes thinks more complicated, specifically answering "should this crypto operation in TLS 1.2 or X.509 or PK be done with PSA APIs or legacy APIs?".

Currently the answer is a simple "if and only if `USE_PSA_CRYPTO` is enabled" (if we disregard restartable ECC for argument's sake - but even then, the decision is made at compile time). If we generalize `pk_setup_opaque()` and `ssl_xxx_psk_opaque()` to be present with only `_C` I'm afraid the answer is going to be more complicated: "is `USE_PSA_CRYPTO` enabled? if not, is the core related to an opaque thing in a way that we should still use PSA anyway?" Not sure we want that.

(Note: writing this before I took the time to look at 8910 or investigate about 8905. Perhaps after I do I'll realise things turn out to be simpler than I think. I'm still thinking this looks like non-trivial effort and potential for bad surprises, for something of very little value.)

File: library/pk.c
Comment (valeriosetti) [reply library/pk.c:325]: I removed the comment because it was no more relevant IMO: the entire function is now guarded by MBEDTLS_PSA_CRYPTO_CLIENT which, according to the original comment, it means that "this function relies on a crypto implementation which is Mbed TLS compliant and therefore has also enrollment functions".
But perhaps I'm missing something like -> MBEDTLS_PSA_CRYPTO_CLIENT: there is a crypto provider, but it might be not Mbed TLS compliant?

Comment (valeriosetti) [reply library/pk.c:325]: > Unofficially, we've agreed to help TF-M with using MBEDTLS_PSA_CRYPTO_CLIENT with alternative implementations of the PSA Crypto specification.

Ok, thanks for the explanation. So what about guarding both calls (set/get) to enrollment functions with `#if defined(MBEDTLS_PSA_CRYPTO_C)`?
When doing our usual tests we'll have `CRYPTO_C` and therefore `CRYPTO_CLIENT` so basically nothing changes. On TFM side they will have `CRYPTO_CLIENT` and `!CRYPTO_C` so they won't need to deal with those functions.

Comment (gilles-peskine-arm) [reply library/pk.c:325]: At this stage and given we're in a hurry I propose to keep calls to the get function as they are. In addition to the one here, there's one in `mbedtls_pk_sign_ext`.

Comment (gilles-peskine-arm) [reply library/pk.c:325]: > action on me in `mbedtls_pk_get_psa_attributes`

Actually, currently, in `development`, the call to `psa_set_key_enrollment_algorithm` is guarded by `#if defined(MBEDTLS_PSA_CRYPTO_C)`, since that's the condition for `mbedtls_pk_get_psa_attributes`. So actually, the action here is: in this pull request, when generalizing `mbedtls_pk_get_psa_attributes` to `CLIENT`, please keep a `C` guard around the call to `psa_set_key_enrollment_algorithm`, with a comment to explain why.

**Comment by valeriosetti** [comment]
- Rebased on #8774 (shrinked in the [1st commit](https://github.com/Mbed-TLS/mbedtls/pull/8842/commits/a98f844c28c417ed6acdd39dbb3169d09efb340c))
- Addressed all the comments
- Added @gilles-peskine-arm commits from #8910

File: ChangeLog.d/8710.txt
Comment (gilles-peskine-arm) [reply ChangeLog.d/8710.txt]: > Who wants to use PSA-held keys with TLS but also that TLS use the legacy API for everything else? 

A number of people don't use `MBEDTLS_USE_PSA_CRYPTO`:

* People who use Linux distributions (or BSD ports, etc.) built with the default configuration.
* People who see that we've fixed some pretty bad bugs with `MBEDTLS_USE_PSA_CRYPTO` and are worried what else we haven't discovered yet.

Some of those people are still interested in secure element support in TLS. Admittedly, while PSA opaque keys are the most natural way to achieve this, there are other ways to do it for signature-based key exchanges (such as `MBEDTLS_SSL_ASYNC_PRIVATE`).

> especially for the PSK case with XXX-PSK key exchanges

PSK is not often used with the constraint “must work with the default configuration”, so that's less important.

> I don't think we want to rush this in 3.6.0

That I agree with. I'm at the point where I'm still hesitating how to best get the `CLIENT` extension: redo this PR and review it from scratch, or just rebase it and get the OPAQUE extension as well?

Comment (gilles-peskine-arm) [reply ChangeLog.d/8710.txt]: Following our discussion on Slack, this discussion is resolved as far as I'm concerned: you're not confident that we can have something reasonable in 3.6.0, so we aren't generalizing `MBEDTLS_USE_PSA_CRYPTO` opaque.

The question is then whether to remove the USE generalization from this PR, or to make a new PR with a clean commit history that only does the C→CLIENT generalization. We've decided that a clean commit history will make the review easier.

**Comment by gilles-peskine-arm** [comment]
Following our discussions [here](https://github.com/Mbed-TLS/mbedtls/pull/8842#discussion_r1516176998) and on Slack, we are abandoning https://github.com/Mbed-TLS/mbedtls/issues/8710 as an objective for 3.6. We do want to keep the objective of generalizing functions guarded by `MBEDTLS_PSA_CRYPTO To facilitate review, we are going to do that in a new pull request with a clean history, rather than have do-undo in this pull request.

Actions:

* [Valerio] https://github.com/Mbed-TLS/mbedtls/issues/8919 Make a new PR with a clean commit history that only does the C→CLIENT generalization.
* [Gilles] Update https://github.com/Mbed-TLS/mbedtls/issues/8714 if needed
* [can wait until after the release] https://github.com/Mbed-TLS/mbedtls/issues/8712

**Pull request closed without merging**
